[She Peiran] iP

s-peiran opened the pull request:
#DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://user-images.githubusercontent.com/25423296/163456776-7f95b81a-f1ed-45f7-b7ab-8fa810d529fa.png">
  <source media="(prefers-color-scheme: light)" srcset="https://user-images.githubusercontent.com/25423296/163456779-a8556205-d0a5-45e2-ac17-42d089e3c3f8.png">
  <img alt="Shows an illustrated sun in light mode and a moon with stars in dark mode." src="https://user-images.githubusercontent.com/25423296/163456779-a8556205-d0a5-45e2-ac17-42d089e3c3f8.png">
</picture>

- text-based
* easy to learn
+ ~~FAST~~ SUPER FAST to use 🤠 

All you need to do is,

1. download it from [here](https://github.com/s-peiran/ip/releases).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you :

And it is __FREE__!

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)
- [ ] If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

s-peiran committed:
level 0

s-peiran committed:
level 1

s-peiran committed:
update Level-2

s-peiran committed:
Level-3

s-peiran committed:
complete Level-4

s-peiran committed:
A-TextUiTesting

s-peiran committed:
Level-5

s-peiran committed:
Level-6 with Exception enchancements

s-peiran committed:
add save function

s-peiran committed:
commit to master branch

s-peiran committed:
commit for merging

s-peiran committed:
modify deadline

s-peiran committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/functional/DukeException.java
#	src/main/java/tasks/Deadline.java

s-peiran committed:
add dateTime to Deadline

s-peiran committed:
progress on master branch

s-peiran committed:
add dateTime for Event

s-peiran committed:
add event for master branch

s-peiran committed:
Merge branch 'Level-8'

# Conflicts:
#	src/main/java/tasks/Deadline.java

s-peiran committed:
Merge branch 'add-gradle-support'

Daphne789 commented on src/main/java/tasks/Event.java:
```diff
@@ -1,30 +1,45 @@
 package tasks;
 
+import java.time.DateTimeException;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
```
```suggestion
import java.time.LocalDate;

import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.time.DateTimeException;
```
Should it be ordered logically into the relevant group?

Daphne789 commented on src/main/java/tasks/Event.java:
```diff
@@ -1,30 +1,45 @@
 public class Event extends Task {
     String[] split1;
     String[] splitStart;
     String[] splitEnd;
```
Perhaps can use more intuitive variable names?

Daphne789 commented on src/main/java/tasks/Deadline.java:
```diff
@@ -8,26 +8,20 @@
             try {
                 DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d/MM/yyyy");
-                this.dueDate = LocalDate.parse(split2[1], formatter);
+                this.dueDate = LocalDate.parse(split2[1], formatter).format(DateTimeFormatter.ofPattern("MMM d yyyy"));
```
```suggestion
                this.dueDate = LocalDate.parse(split2[1], formatter)
                    .format(DateTimeFormatter.ofPattern("MMM d yyyy"));
```

Daphne789 commented on src/main/java/commands/UnmarkCommand.java:
```diff
@@ -0,0 +1,25 @@
+        Task job = (Task) tasks.get(index);
+        job = job.unmark();
+        tasks.set(index, job);
+        System.out.println("____________________________________________________________\n" +
+                "OK, I've marked this task as not done yet:\n" + job.toString() + "\n" +
+                "____________________________________________________________");
```
Should the implementation of the horizontal line be extracted to make it simpler?

Daphne789 commented on src/main/java/commands/Command.java:
```diff
@@ -0,0 +1,22 @@
+
+
+public class Command {
+    boolean isExit;
```
```suggestion
    boolean hasExit;
```

Daphne789 commented on src/main/java/functional/Parser.java:
```diff
@@ -0,0 +1,25 @@
+
+    public static Command parse(String fullCommand) {
+        String[] input = fullCommand.split(" ");
+        switch (input[0]) {
+            case "bye":
+                return new ExitCommand();
+            case "delete":
+                return new DeleteCommand();
+            case "list":
+                return new ListCommand();
+            case "mark":
+                return new MarkCommand();
+            case "unmark":
+                return new UnmarkCommand();
+            default:
+                return new AddCommand();
```
```suggestion
        switch (input[0]) {
        case "bye":
            return new ExitCommand();
        case "delete":
            return new DeleteCommand();
        case "list":
            return new ListCommand();
        case "mark":
            return new MarkCommand();
        case "unmark":
            return new UnmarkCommand();
        default:
            return new AddCommand();
```

NgChunMan replied:
Perhaps no indentation for case clauses would be better as it is part of the coding standard?

Daphne789 reviewed (commented):
I found your code easy to read for the most part except in a few places
where you do the splitting. I noted a few minor coding standard violations
too.
Overall, Great Job!

s-peiran committed:
test with JUnit

s-peiran committed:
work on A-Jar

NgChunMan commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,72 @@
+
+import java.util.Arrays;
+
+public class AddCommand extends commands.Command {
```
Any reason why you need to extends commands.Command rather than just extends Command?

NgChunMan commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,72 @@
+package commands;
+
+import functional.DukeException;
+import functional.TaskList;
+import functional.Ui;
+import tasks.Deadline;
+import tasks.Event;
+import tasks.Task;
+import tasks.ToDo;
```
I like the way you organise your files in different packages. It's very neat to me!

NgChunMan commented on src/main/java/commands/AddCommand.java:
```diff
@@ -0,0 +1,72 @@
+                Task job = new ToDo(ui.getInput(), marked);
+                tasks.add(job);
+                if (!load) {
+                    System.out.println(job.addTask(tasks.size()));
+                }
+                break;
+            case "deadline":
+                try {
+                    job = new Deadline(ui.getInput(), marked);
+                    tasks.add(job);
+                    if (!load) {
+                        System.out.println(job.addTask(tasks.size()));
+                    }
+                } catch (ArrayIndexOutOfBoundsException e) {
+                    System.out.println("____________________________________________________________\n" +
+                            "☹ OOPS!!! I'm sorry, the format entered was incorrect :-(\n" +
+                            "____________________________________________________________");
+                    throw new DukeException();
+                }
+                break;
+            case "event":
+                try {
+                    job = new Event(ui.getInput(), marked);
+                    tasks.add(job);
+                    if (!load) {
+                        System.out.println(job.addTask(tasks.size()));
+                    }
+                } catch (ArrayIndexOutOfBoundsException e) {
+                    System.out.println("____________________________________________________________\n" +
+                            "☹ OOPS!!! I'm sorry, the format entered was incorrect :-(\n" +
+                            "____________________________________________________________");
+                    throw new DukeException();
+                }
+                break;
+            default:
+                System.out.println("____________________________________________________________\n" +
+                        "☹ OOPS!!! I'm sorry, but I don't know what that means :-(\n" +
+                        "____________________________________________________________");
+                throw new DukeException();
+        }
```
Perhaps no indentation for case clauses would be better as it is part of the coding standard?

NgChunMan commented on src/main/java/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,27 @@
+import tasks.Task;
+
+
+public class DeleteCommand extends commands.Command {
```
Any reason why you need to extends commands.Command rather than just extends Command?

NgChunMan commented on src/main/java/commands/MarkCommand.java:
```diff
@@ -0,0 +1,25 @@
+import functional.Ui;
+import tasks.Task;
+
+public class MarkCommand extends commands.Command {
```
... I noticed the same issue in several other places too.

NgChunMan commented on src/main/java/commands/MarkCommand.java:
```diff
@@ -0,0 +1,25 @@
+        super();
+    }
+
+    public void execute(TaskList<Task> tasks, Ui ui, boolean marked, boolean load) throws functional.DukeException {
```
Any reason why you throws functional.DukeException rather than just throws DukeException?

NgChunMan commented on src/main/java/commands/UnmarkCommand.java:
```diff
@@ -0,0 +1,25 @@
+import functional.Ui;
+import tasks.Task;
+
+public class UnmarkCommand extends commands.Command {
```
... I noticed the same issue in several other places too.

NgChunMan commented on src/main/java/functional/Storage.java:
```diff
@@ -0,0 +1,61 @@
+package functional;
+
+import commands.Command;
+import tasks.Task;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.List;
+import java.util.Scanner;
```
Good practice on organising the import statements!

NgChunMan commented on src/main/java/functional/Ui.java:
```diff
@@ -0,0 +1,65 @@
+                "____________________________________________________________");
+    }
+
+    public String readCommand() throws functional.DukeException {
```
... I noticed the same issue in several other places too

NgChunMan commented on src/main/java/functional/Ui.java:
```diff
@@ -0,0 +1,65 @@
+        System.out.println("No preexisting file\n" + "creating new file");
+    }
+
+    public static void main(String[] args) {
```
Any reason why you need a main function in Ui class?

NgChunMan reviewed (commented):
LGTM. I found your code easy to read for the most part but I noted a few minor coding standard violations too. Just a few nits to fix. Remember to add back the Javadoc comments for the methods too.

s-peiran committed:
add JavaDoc

s-peiran committed:
tweak the code to comply with coding standard

s-peiran committed:
implement find

s-peiran committed:
advance on master

s-peiran committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/commands/AddCommand.java

s-peiran committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/commands/AddCommand.java
#	src/main/java/commands/Command.java
#	src/main/java/commands/DeleteCommand.java
#	src/main/java/commands/MarkCommand.java
#	src/main/java/commands/UnmarkCommand.java
#	src/main/java/functional/Storage.java
#	src/main/java/functional/Ui.java

s-peiran committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/commands/ListCommand.java
#	src/main/java/functional/Parser.java

s-peiran committed:
add checkstyle feature

s-peiran committed:
Merge branch 'branch-A-CheckStyle'

s-peiran committed:
implement GUI

s-peiran committed:
Merge branch 'branch-Level-10'

s-peiran committed:
add varargs

s-peiran committed:
Merge branch 'brach-A-Varargs'

s-peiran committed:
Use assertions in husky class

s-peiran committed:
improve code quality

s-peiran committed:
Merge pull request #2 from s-peiran/branch-A-Assertions

Use Assertions

s-peiran committed:
Merge pull request #3 from s-peiran/branch-A-CodeQuality

Improve code quality

s-peiran committed:
Implement sorting of TaskList by chronological order

TaskList order is sorted by order of task creation.

Such order of the TaskList when displayed is not intuitive.

Let's make the Task class a comparable, and call the 'sort' method of collections when the user displays the tasklist.

The TaskList is now displayed in a user friendly and intuitive format.
Sorting the order of the tasks when the user calls list can reduce the number of time 'sort' is called, hence increasing efficiency.

s-peiran committed:
Merge pull request #4 from s-peiran/branch-C-Sort

Implement sorting of TaskList by chronological order

s-peiran committed:
implement recurring tasks

users have to manually add in weekly repeating tasks.

This process of adding the same task every week is tedious.
The recurring task functionality allow users to add repeating tasks only once, and it will be auto updated when past.

Let's check whether recurring tasks are past when user calls 'list'. If the recurring task is past (i.e. expired), we update the date of the task
to the corresponding day of the week after the current day.

We only display one instance of the recurring task on the task list at any point of time to reduce cluttering.

s-peiran committed:
Merge pull request #5 from s-peiran/branch-B-RecurringTasks

implement recurring tasks

s-peiran committed:
Add User Guide

s-peiran committed:
Modify the readme

s-peiran committed:
Add download link to Userguide

s-peiran committed:
Add continuous Integration

s-peiran committed:
Add Javadoc to new methods and ensure checkstyle

s-peiran committed:
Merge branch 'branch-A-CI'

s-peiran committed:
Add more JUnit tests

s-peiran committed:
Fix checkstyle

s-peiran committed:
Delete text-ui-test/runtest.sh

s-peiran committed:
Update runtest.sh to search for subdirectories

s-peiran committed:
Fix runtest.sh path

s-peiran committed:
Update text-ui-test/runtest.sh to locate Husky instead of Duke

s-peiran committed:
Modified runtest.bat to address Illegal char <*> at index 17: ..\src\main\java\*.java

s-peiran committed:
Modified runtest.bat classpath to duke.Husky

s-peiran committed:
Merge pull request #6 from s-peiran/branch-A-MoreTesting

Add more JUnit tests

s-peiran committed:
Update user guide

s-peiran committed:
Fix User Guide

s-peiran committed:
Merge pull request #7 from s-peiran/update-ug

Update user guide

s-peiran committed:
Update README.md
